File: src/tides.js
// NOAA CO-OPS tide predictions for Marshall, Tomales Bay. Marshall is a
// subordinate station, so NOAA publishes only the highs and lows; the curve
// between them is a cosine, the same approximation the printed tide tables
// assume.

import { MIN, yyyymmddUTC } from './time.js';

export const STATION = {
  id: '9415625',
  name: 'Marshall, Tomales Bay',
  lat: 38.1616,
  lng: -122.8886,
  datum: 'MLLW',
  units: 'ft',
};

const API = 'https://api.tidesandcurrents.noaa.gov/api/prod/datagetter';

export function predictionsUrl(begin, end, station = STATION.id) {
  const q = new URLSearchParams({
    product: 'predictions', application: 'MarshallTides',
    begin_date: yyyymmddUTC(begin), end_date: yyyymmddUTC(end),
    datum: STATION.datum, station, time_zone: 'gmt', units: 'english',
    interval: 'hilo', format: 'json',
  });
  return `${API}?${q}`;
}

/** NOAA JSON -> [{time, height, type: 'H'|'L'}] sorted by time. */
export function parsePredictions(json) {
  if (json.error) throw new Error(`NOAA: ${json.error.message}`);
  return json.predictions
    .map(p => ({ time: new Date(p.t.replace(' ', 'T') + 'Z'), height: +p.v, type: p.type }))
    .sort((a, b) => a.time - b.time);
}

export async function fetchExtremes(begin, end, fetchImpl = fetch) {
  const res = await fetchImpl(predictionsUrl(begin, end));
  if (!res.ok) throw new Error(`NOAA ${res.status}`);
  return parsePredictions(await res.json());
}

/** Height at instant t, cosine-interpolated between neighbouring extremes. */
export function heightAt(extremes, t) {
  const ms = +t;
  if (ms <= +extremes[0].time) return extremes[0].height;
  for (let i = 1; i < extremes.length; i++) {
    const a = extremes[i - 1], b = extremes[i];
    if (ms <= +b.time) {
      const f = (ms - a.time) / (b.time - a.time);
      return (a.height + b.height) / 2 + (a.height - b.height) / 2 * Math.cos(Math.PI * f);
    }
  }
  return extremes[extremes.length - 1].height;
}

/** Sampled curve from start to end inclusive. */
export function curve(extremes, start, end, stepMs = 10 * MIN) {
  const pts = [];
  for (let t = +start; t < +end; t += stepMs) pts.push({ time: new Date(t), height: heightAt(extremes, t) });
  pts.push({ time: new Date(end), height: heightAt(extremes, end) });
  return pts;
}

/** Where the tide is right now and what comes next. */
export function tideNow(extremes, now) {
  const i = extremes.findIndex(e => e.time > now);
  const next = i >= 0 ? extremes[i] : null;
  const prev = i > 0 ? extremes[i - 1] : null;
  return {
    height: heightAt(extremes, now),
    rising: next ? next.type === 'H' : false,
    next, prev,
    upcoming: i >= 0 ? extremes.slice(i) : [],
  };
}

File: apple/Shared/Tides.swift
import Foundation

struct TideExtreme: Codable, Equatable {
    let time: Date
    let height: Double
    let isHigh: Bool
    var label: String { isHigh ? "High" : "Low" }
}

/// Where the tide is right now and what comes next.
struct TideState {
    let height: Double
    let rising: Bool
    let next: TideExtreme?
    let previous: TideExtreme?
    let upcoming: [TideExtreme]
}

enum NOAA {
    private static let api = "https://api.tidesandcurrents.noaa.gov/api/prod/datagetter"

    private static let dayFormatter: DateFormatter = {
        let f = DateFormatter()
        f.locale = Locale(identifier: "en_US_POSIX")
        f.timeZone = TimeZone(identifier: "UTC")
        f.dateFormat = "yyyyMMdd"
        return f
    }()

    private static let timeFormatter: DateFormatter = {
        let f = DateFormatter()
        f.locale = Locale(identifier: "en_US_POSIX")
        f.timeZone = TimeZone(identifier: "UTC")
        f.dateFormat = "yyyy-MM-dd HH:mm"
        return f
    }()

    /// Hi/lo predictions in GMT for the station. Marshall is a subordinate
    /// station, so highs and lows are all NOAA publishes for it.
    static func predictionsURL(begin: Date, end: Date, station: String = Station.id) -> URL {
        var c = URLComponents(string: api)!
        c.queryItems = [
            .init(name: "product", value: "predictions"),
            .init(name: "application", value: "MarshallTides"),
            .init(name: "begin_date", value: dayFormatter.string(from: begin)),
            .init(name: "end_date", value: dayFormatter.string(from: end)),
            .init(name: "datum", value: "MLLW"),
            .init(name: "station", value: station),
            .init(name: "time_zone", value: "gmt"),
            .init(name: "units", value: "english"),
            .init(name: "interval", value: "hilo"),
            .init(name: "format", value: "json"),
        ]
        return c.url!
    }

    private struct Response: Decodable {
        struct Prediction: Decodable { let t: String; let v: String; let type: String }
        struct APIError: Decodable { let message: String }
        let predictions: [Prediction]?
        let error: APIError?
    }

    struct ParseError: Error, LocalizedError {
        let message: String
        var errorDescription: String? { message }
    }

    static func parse(_ data: Data) throws -> [TideExtreme] {
        let r = try JSONDecoder().decode(Response.self, from: data)
        if let e = r.error { throw ParseError(message: "NOAA: \(e.message)") }
        guard let preds = r.predictions else { throw ParseError(message: "NOAA: no predictions") }
        return try preds.map { p in
            guard let t = timeFormatter.date(from: p.t), let v = Double(p.v) else {
                throw ParseError(message: "NOAA: bad row \(p.t) \(p.v)")
            }
            return TideExtreme(time: t, height: v, isHigh: p.type == "H")
        }.sorted { $0.time < $1.time }
    }
}

enum Tides {
    /// Height at `t`, cosine-interpolated between the neighbouring extremes,
    /// the same approximation the printed tide tables assume.
    static func height(at t: Date, in extremes: [TideExtreme]) -> Double {
        guard let first = extremes.first, let last = extremes.last else { return 0 }
        if t <= first.time { return first.height }
        for i in 1..<extremes.count {
            let a = extremes[i - 1], b = extremes[i]
            if t <= b.time {
                let f = t.timeIntervalSince(a.time) / b.time.timeIntervalSince(a.time)
                return (a.height + b.height) / 2 + (a.height - b.height) / 2 * cos(.pi * f)
            }
        }
        return last.height
    }

    struct Point { let time: Date; let height: Double }

    static func curve(_ extremes: [TideExtreme], from start: Date, to end: Date, step: TimeInterval = 6 * 60) -> [Point] {
        var pts: [Point] = []
        var t = start
        while t < end {
            pts.append(Point(time: t, height: height(at: t, in: extremes)))
            t = t.addingTimeInterval(step)
        }
        pts.append(Point(time: end, height: height(at: end, in: extremes)))
        return pts
    }

    static func state(at now: Date, in extremes: [TideExtreme]) -> TideState {
        let i = extremes.firstIndex { $0.time > now }
        let next = i.map { extremes[$0] }
        let previous = i.flatMap { $0 > 0 ? extremes[$0 - 1] : nil }
        return TideState(height: height(at: now, in: extremes),
                         rising: next?.isHigh ?? false,
                         next: next, previous: previous,
                         upcoming: i.map { Array(extremes[$0...]) } ?? [])
    }
}
